internal/services/organisation.go

```diff
@@ -1,95 +1,122 @@
-//
-//		}
-//	}()
-//
-//	response := &types.Organisation{
-//		OrganisationID:     organisation.ID,
-//		OrganisationName:   organisation.OrganisationName,
-//		LogoURL:            *organisation.LogoURL,
-//		Website:            *organisation.Website,
-//		OrganisationSize:   *organisation.OrganisationSize,
-//		Description:        organisation.Description,
-//		RegistrationNumber: *organisation.RegistrationNumber,
-//		FoundingDate:       organisation.FoundingDate,
-//	}
-//
-//	if err = uw.Commit(tx); err != nil {
-//		return nil, errors.New("organisation could not be created")
-//	}
-//
-//	return response, nil
-//}
+func (d *DefaultOrganisationService) CreateOrganisation(body *types.CreateOrgReq, logger *log.Entry) error {
+	uw := repository.NewGormUnitOfWork(d.db)
+	tx, err := uw.Begin()
+
+	defer func() {
+		if err != nil {
+			logger.Error(err.Error())
+			tx.Rollback()
+		}
+	}()
+
+	if err != nil {
+		logger.Error(err.Error())
+		return err
+	}
+	user := &model.User{
+		Base: model.Base{
+			ID: body.UserID,
+		},
+	}
+	d.db.First(&user)
+	if user.Tier != 1 {
+		logger.Error("Upgrade your acount to create Organisation")
+		return errors.New("Upgrade your acount to create Organisation")
+	}
+	founding, err := dateparse.ParseLocal(body.FoundingDate)
+
+	if err != nil {
+		logger.Error(fmt.Sprintf("Founding date is not a valid date. %s %s", body.FoundingDate, err.Error()))
+		return errors.New(constant.InternalServerError)
+	}
+
+	body.OrganisationName = utils.CapitalizeFirstCharacter(body.OrganisationName)
+
+	organisation := &model.Organisation{}
+
+	filter := map[string]interface{}{"organisation_name": body.OrganisationName}
+	err = d.repo.WithTx(tx).Find(filter, organisation)
+	if err != nil {
+		logger.Error(err.Error())
+		return errors.New(constant.InternalServerError)
+	}
+	if organisation.Base.ID != "" {
+		logger.Error(err.Error())
+		return errors.New("organisation name already exist")
+	}
+
+	organisation.UserID = user.ID
+	organisation.OrganisationName = body.OrganisationName
+	organisation.Description = utils.FormatDate(founding)
+	organisation.OrganisationSize = body.OrganisationSize
+
+	err = d.repo.WithTx(tx).Create(organisation)
+	if err != nil {
+		logger.Error(err.Error())
+		return errors.New(constant.InternalServerError)
+	}
+	branch := &model.Branch{}
+
+	branch.OrganisationID = organisation.Base.ID
+	branch.BranchName = organisation.OrganisationName
+	branch.IsHQ = true
+	er := d.rep.WithTx(tx).Create(branch)
+
+	if er != nil {
+		logger.Error(err.Error())
+		return errors.New(constant.InternalServerError)
+	}
+	// Activity log
+	activity := &model.Activity{For: user.ID, Message: fmt.Sprintf("Created an organisation %s", organisation.OrganisationName)}
+	go func() {
+		err := d.activity.Create(activity)
+		if err != nil {
+			logger.Error(err.Error())
+
+		}
+	}()
+	if err = uw.Commit(tx); err != nil {
+		logger.Error(err.Error())
+		return errors.New("organisation could not be created")
+	}
+	return nil
```

Comment:
This function should return information about the created Organisation ie. ID, Name, IsHQ